phase1/gce/configure-vm-kubeadm.sh

```diff
@@ -3,30 +3,52 @@
 TOKEN=$(get_metadata "k8s-kubeadm-token")
 KUBEADM_VERSION=$(get_metadata "k8s-kubeadm-version")
 KUBERNETES_VERSION=$(get_metadata "k8s-kubernetes-version")
-
+KUBELET_VERSION=$(get_metadata "k8s-kubelet-version")
 curl -s https://packages.cloud.google.com/apt/doc/apt-key.gpg | apt-key add -
 
-if [[ "${KUBEADM_VERSION}" == stable ]]; then
+if [[ "${KUBELET_VERSION}" == stable ]]; then
   cat <<EOF > /etc/apt/sources.list.d/kubernetes.list
   deb http://apt.kubernetes.io/ kubernetes-xenial main
 EOF
-
   apt-get update
+  # kubeadm is installed with the kubelet so that the 
+  # kubelet has the configuration at a matching version
   apt-get install -y kubelet kubeadm kubectl kubernetes-cni
-elif [[ "${KUBEADM_VERSION}" == "gs://"* ]]; then
+elif [[ "${KUBELET_VERSION}" == "gs://"* ]]; then
   TMPDIR=/tmp/k8s-debs
   mkdir $TMPDIR
-  gsutil rsync "${KUBEADM_VERSION}" $TMPDIR
+  gsutil rsync "${KUBELET_VERSION}" $TMPDIR
+  # kubeadm is installed with the kubelet so that the 
+  # kubelet has the configuration at a matching version
   dpkg -i $TMPDIR/{kubelet,kubeadm,kubectl,kubernetes-cni}.deb || echo Ignoring expected dpkg failure
   apt-get install -f -y
   systemctl enable kubelet
   systemctl start kubelet
   rm -rf $TMPDIR
 else
-  echo "Don't know how to handle version: $KUBEADM_VERSION"
+  echo "Don't know how to handle kubelet version: $KUBELET_VERSION"
   exit 1
 fi
 
+if [[ "${KUBEADM_VERSION}" != "${KUBELET_VERSION}" ]]; then
+  if [[ "${KUBEADM_VERSION}" == stable ]]; then
+    # Cannot install packages as they will update the 
```

Comment:
I'm confused by this. Can't we still do `apt-get install kubeadm` and `dpkg -i kubelet.deb` separately, from different sources?

Reply:
The trouble is if you install kubeadm via a package, it will update the kubelet config. This can break the kubelet if it is restarted. If we want kubeadm and kubelet at different versions, we need to not use the package installation method to avoid the kubelet config update.

Eg. 
Version N Kubelet using Version N Config Installed by Version N Kubeadm Package
Version N Kubelet using Version N+1 Config installed by Version N+1 Kubeadm Package
N+1 config can contain new flags that version N Kubelet does not recognize etc.

Reply:
Ah, got it. So this is going to necessarily be a hack because the entire way we package and handle the `10-kubeadm.conf` file breaks us, by having a static set of kubelet flags installed from the kubeadm package. I'll take another look with this in mind.

Thanks for clearing that up.

Reply:
It sounds like this is going to break our manual upgrade instructions, too, if kubelet is restarted at the wrong time. Do we have someone owning getting those figured out?